### PR TITLE
feat: upgrade react

### PR DESCRIPTION
**Pull request by IrakliJani**
### Description:

- upgrades to react 18 (https://react.dev/blog/2022/03/08/react-18-upgrade-guide) + (internal doc about react 18 https://www.notion.so/lightdash/2024-Frontend-Playbook-d6c7b355cf8747db85366177b007ab58?pvs=4#e983553d3cc3409882a9d8c02bb609ec)
  - breaking changes:
   - type `FC` does not include `children` in the definition so you have to explicitly pair it with `FC<PropsWithChildren<YourPropsHere>>`
- upgrades `react-query 3` to `react-query 4` (now its callede `@tanstack/react-query` (https://tanstack.com/query/latest/docs/react/guides/migrating-to-react-query-4)
  - breaking changes:
    - `isLoading` prop does not the same way if useQuery has an option of `{ enabled: false }`
    - query keys should only be arrays `[]` across all of the functions
    - undefined values are not accepted as keys anymore
    - undefined values are not accepted as response values anymore
- upgrades cypress to version 13
- updates e2e tests

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/frontend/src/components/Explorer/ResultsCard/ExplorerResults.tsx [hunk 1]
```diff
@@ -82,7 +83,7 @@ export const ExplorerResults = memo(() => {
     }, [exploreData, additionalMetrics, tableCalculations]);
 
     const cellContextMenu = useCallback(
-        (props) => (
+        (props: any) => (
```
Comment (joaoviana) [inline packages/frontend/src/components/Explorer/ResultsCard/ExplorerResults.tsx:86]: check this any

File: packages/frontend/src/components/common/TagInput/TagInput.tsx [hunk 2]
```diff
@@ -51,7 +51,7 @@ export interface TagInputProps
     onChange?(value: string[]): void;
 
     /** Component used to render values */
-    valueComponent?: React.FC<any>;
+    valueComponent?: React.FC<React.PropsWithChildren<any>>;
```
Comment (joaoviana) [inline packages/frontend/src/components/common/TagInput/TagInput.tsx:54]: can we improve this type?

File: packages/frontend/src/components/common/modal/DeleteChartTileThatBelongsToDashboardModal.tsx [hunk 3]
```diff
@@ -16,11 +16,9 @@ interface Props extends ModalProps {
...
 }
 
-const DeleteChartTileThatBelongsToDashboardModal: FC<Props> = ({
-    name,
-    onConfirm,
-    ...modelProps
-}) => (
+const DeleteChartTileThatBelongsToDashboardModal: FC<
+    React.PropsWithChildren<Props>
+> = ({ name, onConfirm, ...modelProps }) => (
```
Comment (joaoviana) [inline packages/frontend/src/components/common/modal/DeleteChartTileThatBelongsToDashboardModal.tsx:21]: rename to `modalProps`

File: packages/frontend/src/features/scheduler/components/SchedulerModalContent.tsx [hunk 4]
```diff
@@ -217,7 +220,7 @@ const UpdateStateContent: FC<{
                 confirmText="Save"
                 onBack={onBack}
                 onSendNow={handleSendNow}
-                loading={mutation.isLoading || scheduler.isLoading}
+                loading={mutation.isLoading || scheduler.isInitialLoading}
```
Comment (joaoviana) [inline packages/frontend/src/features/scheduler/components/SchedulerModalContent.tsx:223]: check if `mutation.isLoading` is correct

File: packages/frontend/src/hooks/dashboard/useDashboard.ts [hunk 5]
```diff
@@ -293,7 +297,8 @@ export const useCreateMutation = (
                         : undefined,
                 });
 
-                useQueryOptions?.onSuccess?.(result);
+                // TODO: fixme...
+                // useQueryOptions?.onSuccess?.(result);
```
Comment (joaoviana) [inline packages/frontend/src/hooks/dashboard/useDashboard.ts:301]: revisit

File: packages/frontend/src/providers/ExplorerProvider.tsx [hunk 6]
```diff
@@ -1131,14 +1131,16 @@ type ConfigCacheMap = {
...
-        typeof useQueryResults | typeof useChartVersionResultsMutation
-    >;
-}> = ({
+export const ExplorerProvider: FC<
+    React.PropsWithChildren<{
+        isEditMode?: boolean;
+        initialState?: ExplorerReduceState;
+        savedChart?: SavedChart;
+        queryResults: ReturnType<
+            typeof useQueryResults | typeof useChartVersionResultsMutation
```
Comment (joaoviana) [inline packages/frontend/src/providers/ExplorerProvider.tsx:1140]: check that `useQueryResults` has the right import

File: packages/frontend/src/providers/__mocks__/TrackingProvider.tsx [hunk 7]
```diff
@@ -1,17 +1,21 @@
...
+}) => {
     return <>{children}</>;
 };
 
 export function useTracking() {
     return jest.fn();
 }
 
-export const TrackPage: FC<any> = ({ children }) => {
+export const TrackPage: FC<React.PropsWithChildren<any>> = ({ children }) => {
```
Comment (joaoviana) [inline packages/frontend/src/providers/__mocks__/TrackingProvider.tsx:13]: check if we need `<any>` in this `TrackPage`, `TrackSection`, and `TrackingProvider`

**Review by joaoviana — COMMENTED** [review]
first review: done 👍

File: packages/frontend/src/components/common/TagInput/TagInput.tsx (hunk 2 above)
Comment (joaoviana) [reply packages/frontend/src/components/common/TagInput/TagInput.tsx:54]: Nope, it's intentional

File: packages/frontend/src/components/common/modal/DeleteChartTileThatBelongsToDashboardModal.tsx (hunk 3 above)
Comment (joaoviana) [reply packages/frontend/src/components/common/modal/DeleteChartTileThatBelongsToDashboardModal.tsx:21]: done

File: packages/frontend/src/hooks/dashboard/useDashboard.ts (hunk 5 above)
Comment (joaoviana) [reply packages/frontend/src/hooks/dashboard/useDashboard.ts:301]: can safely removed, not being used!

File: packages/frontend/src/providers/ExplorerProvider.tsx (hunk 6 above)
Comment (joaoviana) [reply packages/frontend/src/providers/ExplorerProvider.tsx:1140]: it does, it's from a local file

File: packages/frontend/src/providers/__mocks__/TrackingProvider.tsx (hunk 7 above)
Comment (joaoviana) [reply packages/frontend/src/providers/__mocks__/TrackingProvider.tsx:13]: we dont

**Comment by joaoviana** [comment]
TODO: write description

File: packages/frontend/src/features/scheduler/components/SchedulerModalContent.tsx (hunk 4 above)
Comment (IrakliJani) [reply packages/frontend/src/features/scheduler/components/SchedulerModalContent.tsx:223]: @joaoviana  mutation states are the same, they don't have `isInitialloading` so no need to update it

Comment (joaoviana) [reply packages/frontend/src/features/scheduler/components/SchedulerModalContent.tsx:223]: amazing!

File: packages/frontend/src/hooks/useActiveProject.ts [hunk 8]
```diff
@@ -1,23 +1,19 @@
...
+import { useEffect } from 'react';
 import { useParams } from 'react-router-dom';
-import { UseQueryFetchOptions } from '../types/UseQuery';
 import { useDefaultProject, useProjects } from './useProjects';
 
 const LAST_PROJECT_KEY = 'lastProject';
 
 export const useActiveProject = () => {
-    return useQuery<string | undefined>(
+    return useQuery<string | null>(
```
Comment (IrakliJani) [inline packages/frontend/src/hooks/useActiveProject.ts:14]: useQuery no longer accepts undefined as a response type

File: packages/frontend/src/hooks/dashboard/useDashboard.ts [hunk 9]
```diff
@@ -292,8 +294,6 @@ export const useCreateMutation = (
                           }
                         : undefined,
                 });
-
-                useQueryOptions?.onSuccess?.(result);
```
Comment (IrakliJani) [inline packages/frontend/src/hooks/dashboard/useDashboard.ts:296]: this option was no longer necessary

File: packages/frontend/src/features/sync/components/SyncModalForm.tsx [hunk 10]
```diff
@@ -109,13 +109,14 @@ export const SyncModalForm: FC<{ chartUuid: string }> = ({ chartUuid }) => {
 
     const hasSetGoogleSheet = methods.watch('options.gdriveId') !== '';
 
-    if (isEditing && (isLoadingSchedulerData || isSchedulerError)) {
-        return isLoadingSchedulerData ? (
-            <SuboptimalState title="Loading Sync" loading />
-        ) : (
-            <ErrorState error={schedulerError.error} />
-        );
+    if (isEditing && isLoadingSchedulerData) {
```
Comment (IrakliJani) [inline packages/frontend/src/features/sync/components/SyncModalForm.tsx:112]: double check

File: packages/frontend/src/features/sync/components/SyncModalDelete.tsx [hunk 11]
```diff
@@ -26,12 +26,12 @@ export const SyncModalDelete = () => {
...
-        return scheduler.isLoading ? (
-            <SuboptimalState title="Loading sync" loading />
-        ) : (
-            <ErrorState error={scheduler.error.error} />
-        );
+    if (scheduler.isInitialLoading) {
+        return <SuboptimalState title="Loading sync" loading />;
+    }
+
+    if (scheduler.error) {
```
Comment (IrakliJani) [inline packages/frontend/src/features/sync/components/SyncModalDelete.tsx:33]: double check

File: packages/e2e/cypress.config.ts [hunk 12]
```diff
@@ -20,7 +20,6 @@ export default defineConfig({
             'analytics.lightdash.com',
         ],
         trashAssetsBeforeRuns: true,
-        videoUploadOnPasses: false,
```
Comment (IrakliJani) [inline packages/e2e/cypress.config.ts:23]: double check

File: packages/frontend/src/features/scheduler/components/SchedulerModalContent.tsx [hunk 13]
```diff
@@ -176,18 +173,18 @@ const UpdateStateContent: FC<{
         [scheduler.data, user?.userUuid, track, sendNow],
     );
 
-    if (scheduler.isLoading || scheduler.error) {
+    if (scheduler.isInitialLoading || scheduler.error) {
         return (
             <>
                 <Box m="xl">
-                    {scheduler.isLoading ? (
+                    {scheduler.isInitialLoading ? (
```
Comment (IrakliJani) [inline packages/frontend/src/features/scheduler/components/SchedulerModalContent.tsx:180]: double check

File: packages/frontend/src/features/scheduler/components/SchedulerDeleteModal.tsx [hunk 14]
```diff
@@ -54,17 +54,13 @@ export const SchedulerDeleteModal: FC<DashboardDeleteModalProps> = ({
...
-                        ) : (
-                            <ErrorState error={scheduler.error.error} />
-                        )}
-                    </>
+                {scheduler.isInitialLoading ? (
+                    <Stack h={300} w="100%" align="center">
+                        <Text fw={600}>Loading schedulers</Text>
+                        <Loader />
+                    </Stack>
+                ) : scheduler.isError ? (
```
Comment (IrakliJani) [inline packages/frontend/src/features/scheduler/components/SchedulerDeleteModal.tsx:62]: double check

File: packages/e2e/cypress/e2e/app/explore.cy.ts [hunk 15]
```diff
@@ -203,9 +205,6 @@ describe('Explore', () => {
                     // run query
                     cy.get('button').contains('Run query').click();
 
-                    // open chart
-                    cy.findByTestId('Charts-card-expand').click();
```
Comment (IrakliJani) [inline packages/e2e/cypress/e2e/app/explore.cy.ts:207]: double check

File: packages/e2e/cypress/e2e/app/downloadCsv.cy.ts [hunk 16]
```diff
@@ -191,18 +197,17 @@ describe('Download CSV on Explore', () => {
             // open chart
             cy.findByTestId('Charts-card-expand').click();
             // Close results
-            cy.findByTestId('Results-card-expand').click();
```
Comment (IrakliJani) [inline packages/e2e/cypress/e2e/app/downloadCsv.cy.ts:194]: double check

File: packages/e2e/cypress/e2e/app/dates.cy.ts [hunk 17]
```diff
@@ -385,19 +397,21 @@ describe('Date tests', () => {
             `/projects/${SEED_PROJECT.project_uuid}/tables/orders${exploreStateUrlParams}`,
         );
 
-        cy.findByTestId('Filters-card-expand').click();
-        cy.findByTestId('SQL-card-expand').click();
```
Comment (IrakliJani) [inline packages/e2e/cypress/e2e/app/dates.cy.ts:389]: double check

**Comment by IrakliJani** [comment]
I manually reviewed **all** the files and left comments on files that require further verification.

CleanShot 2024-01-12 at 17 12 50@2x

File: packages/e2e/cypress.config.ts (hunk 12 above)
Comment (joaoviana) [reply packages/e2e/cypress.config.ts:23]: don't need it anymore with cypress upgrade

Comment (IrakliJani) [reply packages/e2e/cypress.config.ts:23]: CleanShot 2024-01-12 at 17 33 49@2x
its fine I can see a failed test video here

Comment (IrakliJani) [reply packages/e2e/cypress.config.ts:23]: nope! it's not fine. no videos only screenshots 😫 
CleanShot 2024-01-12 at 17 34 52@2x

Comment (joaoviana) [reply packages/e2e/cypress.config.ts:23]: That's intended though! We removed the option to upload videos a few months(?) ago

Comment (IrakliJani) [reply packages/e2e/cypress.config.ts:23]: fixed: 
CleanShot 2024-01-12 at 18 35 37@2x

Comment (IrakliJani) [reply packages/e2e/cypress.config.ts:23]: @joaoviana I remember it being enabled. no?

File: packages/frontend/src/features/scheduler/components/SchedulerModalContent.tsx (hunk 13 above)
Comment (IrakliJani) [reply packages/frontend/src/features/scheduler/components/SchedulerModalContent.tsx:180]: this is fine ✅

File: packages/e2e/cypress.config.ts (hunk 12 above)
Comment (IrakliJani) [reply packages/e2e/cypress.config.ts:23]: https://lightdash.slack.com/archives/C02GQKJK84Q/p1705071638465379

should be enabled ✅

File: packages/frontend/src/features/sync/components/SyncModalDelete.tsx (hunk 11 above)
Comment (IrakliJani) [reply packages/frontend/src/features/sync/components/SyncModalDelete.tsx:33]: this is fine

File: packages/frontend/src/features/sync/components/SyncModalForm.tsx (hunk 10 above)
Comment (IrakliJani) [reply packages/frontend/src/features/sync/components/SyncModalForm.tsx:112]: this is fine ✅

File: packages/frontend/src/hooks/dashboard/useDashboard.ts (hunk 9 above)
Comment (IrakliJani) [reply packages/frontend/src/hooks/dashboard/useDashboard.ts:296]: this is fine ✅

File: packages/frontend/src/hooks/useActiveProject.ts (hunk 8 above)
Comment (IrakliJani) [reply packages/frontend/src/hooks/useActiveProject.ts:14]: this is fine ✅

File: packages/frontend/src/features/scheduler/components/SchedulerDeleteModal.tsx (hunk 14 above)
Comment (IrakliJani) [reply packages/frontend/src/features/scheduler/components/SchedulerDeleteModal.tsx:62]: this is fine ✅

File: packages/frontend/src/api.ts [hunk 18]
```diff
@@ -53,7 +53,9 @@ export const lightdashApi = async <T extends ApiResponse['results']>({
         .then((d: ApiResponse | ApiError) => {
             switch (d.status) {
                 case 'ok':
-                    return d.results as T;
+                    // make sure we return null instead of undefined
+                    // otherwise react-query will crash
+                    return (d.results ?? null) as T;
```
Comment (IrakliJani) [inline packages/frontend/src/api.ts:58]: ^^^ this

**Review by joaoviana — APPROVED** [review]
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉

**Pull request merged**